lumiwealth_tradier/base.py

```diff
@@ -77,25 +108,43 @@ def request(self, endpoint, params=None, headers=None, data=None, method="get")
-        elif method == "post":
-            r = requests.post(url=f"{self.base_url()}/{endpoint}", params=params, headers=headers, data=data)
-        elif method == "delete":
-            r = requests.delete(url=f"{self.base_url()}/{endpoint}", params=params, data=data, headers=headers)
-        else:
-            raise ValueError(f"Invalid method {method}. Must be one of ['get', 'post', 'delete']")
-
-        # Check for errors in response from Tradier API. 
-        # 502 is a common error code when the API is down for a few seconds so we ignore it too.
-        if r.status_code != 200 and r.status_code != 201 and r.status_code != 502:
-            raise TradierApiError(f"Error: {r.status_code} - {r.text}")
-
-        # Parse the response from the Tradier API.  Sometimes no valid json is returned.
-        try:
-            ret_data = r.json()
-        except json.decoder.JSONDecodeError:
-            ret_data = {}
+        # We want to retry GET requests if the response is empty or if the request fails.
+        # We don't want to retry other methods because they may have side effects (like submitting
+        # extra orders).
+        # We use simple a for loop to catch cases when the request succeeds but there was no data.
+        retry_attempts = DEFAULT_RETRY_ATTEMPTS if method == "get" else 1
+        for retry_attempt in range(retry_attempts):
+            r = None
+            if method == "get":
+                # We use a retry session to handle transient errors and retry the request.
+                r = self.requests_retry_session().get(url=f"{self.base_url()}/{endpoint}",params=params,headers=headers)
+            elif method == "post":
+                r = requests.post(url=f"{self.base_url()}/{endpoint}", params=params, headers=headers, data=data)
+            elif method == "delete":
+                r = requests.delete(url=f"{self.base_url()}/{endpoint}", params=params, data=data, headers=headers)
+            else:
+                raise ValueError(f"Invalid method {method}. Must be one of ['get', 'post', 'delete']")
+
+            # Check for errors in response from Tradier API.
+            # 502 is a common error code when the API is down for a few seconds so we ignore it too.
+            if r.status_code != 200 and r.status_code != 201 and r.status_code != 502:
+                raise TradierApiError(f"Error: {r.status_code} - {r.text}")
+
+            # Parse the response from the Tradier API.  Sometimes no valid json is returned.
+            try:
+                ret_data = r.json()
+            except json.decoder.JSONDecodeError:
+                ret_data = {}
+
+            if required_response_key:
+                if required_response_key in ret_data and ret_data[required_response_key] is not None:
+                    break
+                else:
+                    if retry_attempt == retry_attempts - 1:
+                        logger.info(f"Response from {endpoint} did not contain {required_response_key}.")
```

Comment:
This should raise a TradierApiError exception because it still will not return the actual data the user is requesting.

Reply:
Done.

Reply:
Actually, now that i that, a lot of tests fail because each get call has its own specific checks for the response data and raises LookupErrors if it can't find the data it needs. I did change the logger to error so that there is a logger message before the exception is raised.